File: test_runner.js
const http = require('http');

function runTest(name, files, stdin = "", expectedStatus = "success", checkAi = false) {
    console.log(`Running Test: ${name}...`);

    const payload = JSON.stringify({
        codeFiles: JSON.stringify(files),
        stdin: stdin
    });

    const options = {
        hostname: 'localhost',
        port: 3000,
        path: '/compile',
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'Content-Length': Buffer.byteLength(payload)
        }
    };

    const req = http.request(options, (res) => {
        let data = '';
        res.on('data', (chunk) => { data += chunk; });
        res.on('end', () => {
            try {
                const response = JSON.parse(data);
                if (response.status === expectedStatus) {
                    console.log(`  [PASS] Status matches ${expectedStatus}`);
                    if (expectedStatus === "success") {
                        console.log(`  Output: ${response.output.trim()}`);
                    }
                    if (checkAi) {
                        if (response.ai_suggestion) {
                            console.log(`  [PASS] AI Suggestion received: ${response.ai_suggestion.explanation}`);
                        } else {
                            console.log(`  [FAIL] AI Suggestion missing`);
                        }
                    }
                } else {
                    console.log(`  [FAIL] Expected ${expectedStatus}, got ${response.status}`);
                    console.log(`  Response:`, response);
                }
            } catch (e) {
                console.log(`  [ERROR] Failed to parse response: ${e.message}`);
                console.log(`  Raw Data: ${data}`);
            }
            console.log("-".repeat(30));
        });
    });

    req.on('error', (e) => {
        console.error(`  [ERROR] Request failed: ${e.message}`);
    });

    req.write(payload);
    req.end();
}

// Test 1: Hello World
runTest(
    "Hello World",
    [{ name: "main.cpp", content: "#include <iostream>\nint main() { std::cout << \"Hello World\"; return 0; }" }]
);

// Test 2: Input Handling
// Wait a bit to ensure sequential logging (async nature)
setTimeout(() => {
    runTest(
        "Input Handling",
        [{ name: "main.cpp", content: "#include <iostream>\n#include <string>\nint main() { std::string s; std::cin >> s; std::cout << \"Hello \" << s; return 0; }" }],
        "User"
    );
}, 1000);

// Test 3: Compile Error (AI Trigger)
setTimeout(() => {
    runTest(
        "Compile Error (Missing Semicolon)",
        [{ name: "main.cpp", content: "#include <iostream>\nint main() { std::cout << \"Error\" return 0; }" }],
        "",
        "error",
        true
    );
}, 2000);

// Test 4: Missing Header (AI Trigger)
setTimeout(() => {
    runTest(
        "Compile Error (Missing Header)",
        [{ name: "main.cpp", content: "int main() { std::vector<int> v; return 0; }" }],
        "",
        "error",
        true
    );
}, 4000); // Give enough time for AI mock delay

File: test_runner.py
import requests
import json

BASE_URL = "http://localhost:3000/compile"

def run_test(name, files, stdin="", expected_status="success", check_ai=False):
    print(f"Running Test: {name}...")
    payload = {
        "codeFiles": json.dumps(files),
        "stdin": stdin
    }
    try:
        response = requests.post(BASE_URL, data=payload)
        data = response.json()
        
        if data.get("status") == expected_status:
            print(f"  [PASS] Status matches {expected_status}")
            if expected_status == "success":
                print(f"  Output: {data.get('output').strip()}")
            
            if check_ai:
                if "ai_suggestion" in data:
                    print(f"  [PASS] AI Suggestion received: {data['ai_suggestion']['explanation']}")
                else:
                    print(f"  [FAIL] AI Suggestion missing")
        else:
            print(f"  [FAIL] Expected {expected_status}, got {data.get('status')}")
            print(f"  Response: {data}")

    except Exception as e:
        print(f"  [ERROR] Request failed: {e}")
    print("-" * 30)

# Test 1: Hello World
run_test(
    "Hello World", 
    [{"name": "main.cpp", "content": "#include <iostream>\nint main() { std::cout << \"Hello World\"; return 0; }"}]
)

# Test 2: Input Handling
run_test(
    "Input Handling", 
    [{"name": "main.cpp", "content": "#include <iostream>\n#include <string>\nint main() { std::string s; std::cin >> s; std::cout << \"Hello \" << s; return 0; }"}],
    stdin="User"
)

# Test 3: Compile Error (AI Trigger)
run_test(
    "Compile Error (Missing Semicolon)", 
    [{"name": "main.cpp", "content": "#include <iostream>\nint main() { std::cout << \"Error\" return 0; }"}], # Missing ;
    expected_status="error",
    check_ai=True
)

# Test 4: Missing Header (AI Trigger)
run_test(
    "Compile Error (Missing Header)", 
    [{"name": "main.cpp", "content": "int main() { std::vector<int> v; return 0; }"}], # Missing vector include
    expected_status="error",
    check_ai=True
)
